File: mlx-sys/shim/src/metal.cc
#include "cxx_mlx_shim/metal.h"

#include <stdexcept>
#include <string>
#include <variant>

#include "mlx/backend/metal/metal.h"
#include "mlx/device.h"

namespace cxx_mlx {

void start_capture(rust::Str path) {
  mlx::core::metal::start_capture(std::string(path));
}

void stop_capture() {
  mlx::core::metal::stop_capture();
}

rust::String device_architecture() {
  // mlx upstream relocated metal::device_info() to the device-agnostic
  // mlx::core::device_info(const Device&) entry point in mlx/device.h.
  // The legacy metal::device_info() symbol is no longer exported by
  // installed libmlx.a, so we route through the public API instead.
  const auto& info = mlx::core::device_info(
      mlx::core::Device(mlx::core::Device::gpu));
  auto it = info.find("architecture");
  if (it == info.end()) {
    throw std::runtime_error(
        "mlx::core::device_info(gpu) has no 'architecture' entry");
  }
  if (const auto* s = std::get_if<std::string>(&it->second)) {
    return rust::String(*s);
  }
  throw std::runtime_error(
      "mlx::core::device_info(gpu)['architecture'] is not a std::string");
}

}  // namespace cxx_mlx

File: mlx-sys/src/bridge/metal.rs
//! Bridge for `mlx::core::metal::{start,stop}_capture` (debug/profiling) and
//! `mlx::core::metal::device_info()['architecture']` (device-arch lookup
//! tables — used by ironmlx-p8a-stage9 to pick per-chip kernel tiles).
//!
//! All functions in this bridge can throw on the C++ side (driver /
//! capture-manager failure for capture; missing Metal backend or unexpected
//! variant type for architecture), so they're declared `Result<T>` per the
//! project's "shim-can-throw → cxx Result" rule (`bridge/mod.rs`).

#[cxx::bridge(namespace = "cxx_mlx")]
mod ffi_bridge {
    unsafe extern "C++" {
        include!("cxx_mlx_shim/metal.h");

        /// Open an Xcode-compatible `.gputrace` file at `path` and start
        /// capturing every Metal command submitted on the default device.
        ///
        /// Throws on driver / capture-manager failure: missing Xcode
        /// entitlement (`MTL_CAPTURE_ENABLED=1`), path not writable, capture
        /// already running, or no Metal device available.
        fn start_capture(path: &str) -> Result<()>;

        /// Finalize the in-progress capture. Throws if no capture is active.
        fn stop_capture() -> Result<()>;

        /// Configure the exact MLX metallib loaded by this process.
        fn set_metallib_path(path: &str) -> Result<()>;

        /// Return the Metal device's architecture name (e.g. `"apple_g13s"`
        /// for an M1 Pro 16-core GPU, `"apple_g15p"` for M3 Pro), as
        /// reported by `MTLDevice.architecture.name` and exposed via
        /// `mlx::core::metal::device_info()['architecture']`.
        ///
        /// Throws if the Metal backend isn't available on this system or
        /// the entry's variant is unexpectedly not a string (defensive —
        /// MLX always stores it as `std::string`).
        fn device_architecture() -> Result<String>;
    }
}

pub mod ffi {
    pub use super::ffi_bridge::{
        device_architecture, set_metallib_path, start_capture, stop_capture,
    };
}
